File: cpp/seven_game.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <random>
#include <string>
#include <map>
#include <ctime>

using namespace std;

// Масті карт
enum Suit { HEARTS, DIAMONDS, CLUBS, SPADES };
const string SUIT_NAMES[] = {"♥", "♦", "♣", "♠"};
const string SUIT_NAMES_FULL[] = {"Черви", "Буби", "Хрести", "Піки"};

// Структура карти
struct Card {
    int rank;      // Ранг: 6-14 (6-10, J=11, Q=12, K=13, A=14)
    Suit suit;

    string toString() const {
        string rankStr;
        if (rank <= 10) rankStr = to_string(rank);
        else if (rank == 11) rankStr = "J";
        else if (rank == 12) rankStr = "Q";
        else if (rank == 13) rankStr = "K";
        else if (rank == 14) rankStr = "A";
        return rankStr + SUIT_NAMES[suit];
    }
};

// Клас для колоди карт
class Deck {
private:
    vector<Card> cards;

public:
    Deck() {
        // Створюємо колоду з 36 карт (від 6 до туза)
        for (int suit = HEARTS; suit <= SPADES; suit++) {
            for (int rank = 6; rank <= 14; rank++) {
                cards.push_back({rank, static_cast<Suit>(suit)});
            }
        }
    }

    void shuffle() {
        random_device rd;
        mt19937 g(rd());
        std::shuffle(cards.begin(), cards.end(), g);
    }

    vector<Card> dealCards(int count) {
        vector<Card> dealt;
        for (int i = 0; i < count && !cards.empty(); i++) {
            dealt.push_back(cards.back());
            cards.pop_back();
        }
        return dealt;
    }

    bool isEmpty() const {
        return cards.empty();
    }
};

// Клас гравця
class Player {
protected:
    string name;
    vector<Card> hand;

public:
    Player(const string& n) : name(n) {}

    virtual ~Player() {}

    void addCards(const vector<Card>& cards) {
        hand.insert(hand.end(), cards.begin(), cards.end());
        sortHand();
    }

    void sortHand() {
        sort(hand.begin(), hand.end(), [](const Card& a, const Card& b) {
            if (a.suit != b.suit) return a.suit < b.suit;
            return a.rank < b.rank;
        });
    }

    void showHand() const {
        cout << "\n" << name << " має " << hand.size() << " карт(и):\n";
        for (size_t i = 0; i < hand.size(); i++) {
            cout << i + 1 << ". " << hand[i].toString() << "  ";
            if ((i + 1) % 9 == 0) cout << "\n";
        }
        cout << "\n";
    }

    bool hasCards() const {
        return !hand.empty();
    }

    int getCardCount() const {
        return hand.size();
    }

    string getName() const {
        return name;
    }

    virtual int selectCard(const map<Suit, pair<int, int>>& table, bool canPlaySeven) = 0;

    bool playCard(int index, vector<Card>& played) {
        if (index >= 0 && index < hand.size()) {
            played.push_back(hand[index]);
            hand.erase(hand.begin() + index);
            return true;
        }
        return false;
    }

    vector<int> getValidMoves(const map<Suit, pair<int, int>>& table) const {
        vector<int> valid;

        for (size_t i = 0; i < hand.size(); i++) {
            const Card& card = hand[i];

            // Перевірка чи можна зіграти сімку
            if (card.rank == 7) {
                if (table.find(card.suit) == table.end()) {
                    valid.push_back(i);
                    continue;
                }
            }

            // Перевірка чи можна зіграти іншу карту
            auto it = table.find(card.suit);
            if (it != table.end()) {
                int minRank = it->second.first;
                int maxRank = it->second.second;

                if (card.rank == minRank - 1 || card.rank == maxRank + 1) {
                    valid.push_back(i);
                }
            }
        }

        return valid;
    }
};

// Людський гравець
class HumanPlayer : public Player {
public:
    HumanPlayer(const string& n) : Player(n) {}

    int selectCard(const map<Suit, pair<int, int>>& table, bool canPlaySeven) override {
        showHand();

        vector<int> validMoves = getValidMoves(table);

        if (validMoves.empty()) {
            cout << "У вас немає можливих ходів. Пропускаєте хід.\n";
            return -1;
        }

        cout << "Можливі ходи: ";
        for (int idx : validMoves) {
            cout << (idx + 1) << " ";
        }
        cout << "\n";

        int choice;
        while (true) {
            cout << "Виберіть карту (номер) або 0 для пропуску ходу: ";
            cin >> choice;

            if (choice == 0) return -1;

            choice--;  // Перетворюємо в індекс масиву

            if (find(validMoves.begin(), validMoves.end(), choice) != validMoves.end()) {
                return choice;
            }

            cout << "Неправильний вибір! Спробуйте ще раз.\n";
        }
    }
};

// Комп'ютерний гравець
class ComputerPlayer : public Player {
public:
    ComputerPlayer(const string& n) : Player(n) {}

    int selectCard(const map<Suit, pair<int, int>>& table, bool canPlaySeven) override {
        vector<int> validMoves = getValidMoves(table);

        if (validMoves.empty()) {
            cout << name << " пропускає хід.\n";
            return -1;
        }

        // Проста стратегія: вибираємо випадковий хід з можливих
        random_device rd;
        mt19937 g(rd());
        uniform_int_distribution<> dis(0, validMoves.size() - 1);

        int selectedIdx = validMoves[dis(g)];
        cout << name << " грає карту: " << hand[selectedIdx].toString() << "\n";

        return selectedIdx;
    }
};

// Головний клас гри
class SevenGame {
private:
    vector<Player*> players;
    map<Suit, pair<int, int>> table;  // Для кожної масті: мінімальний та максимальний ранг на столі
    int currentPlayer;
    vector<int> consecutivePasses;

public:
    SevenGame() : currentPlayer(0) {}

    ~SevenGame() {
        for (auto player : players) {
            delete player;
        }
    }

    void addPlayer(Player* player) {
        players.push_back(player);
        consecutivePasses.push_back(0);
    }

    void dealCards() {
        Deck deck;
        deck.shuffle();

        int cardsPerPlayer = 36 / players.size();

        for (auto player : players) {
            player->addCards(deck.dealCards(cardsPerPlayer));
        }
    }

    void showTable() const {
        cout << "\n========== СТІЛ ==========\n";

        if (table.empty()) {
            cout << "Стіл порожній. Грайте сімку для початку!\n";
        } else {
            for (const auto& entry : table) {
                Suit suit = entry.first;
                int minRank = entry.second.first;
                int maxRank = entry.second.second;

                cout << SUIT_NAMES_FULL[suit] << ": ";
                for (int rank = minRank; rank <= maxRank; rank++) {
                    Card temp = {rank, suit};
                    cout << temp.toString() << " ";
                }
                cout << "\n";
            }
        }
        cout << "==========================\n\n";
    }

    bool makeMove() {
        Player* player = players[currentPlayer];

        cout << "\n>>> Хід гравця: " << player->getName() << " ("
             << player->getCardCount() << " карт) <<<\n";

        showTable();

        bool canPlaySeven = true;
        for (const auto& entry : table) {
            if (entry.second.first == 7 && entry.second.second == 7) {
                // Якщо всі масті вже мають сімку
                canPlaySeven = false;
            }
        }

        int selectedCard = player->selectCard(table, canPlaySeven);

        if (selectedCard == -1) {
            consecutivePasses[currentPlayer]++;
            currentPlayer = (currentPlayer + 1) % players.size();
            return true;
        }

        vector<Card> played;
        if (player->playCard(selectedCard, played)) {
            Card card = played[0];

            // Оновлюємо стіл
            if (table.find(card.suit) == table.end()) {
                // Нова масть на столі (має бути сімка)
                table[card.suit] = {card.rank, card.rank};
            } else {
                // Розширюємо діапазон для існуючої масті
                if (card.rank < table[card.suit].first) {
                    table[card.suit].first = card.rank;
                }
                if (card.rank > table[card.suit].second) {
                    table[card.suit].second = card.rank;
                }
            }

            consecutivePasses[currentPlayer] = 0;

            cout << player->getName() << " зіграв карту: " << card.toString() << "\n";

            // Перевірка на перемогу
            if (!player->hasCards()) {
                cout << "\n🎉 " << player->getName() << " ВИГРАВ! 🎉\n";
                return false;
            }
        }

        currentPlayer = (currentPlayer + 1) % players.size();
        return true;
    }

    bool allPlayersPassed() const {
        for (int passes : consecutivePasses) {
            if (passes == 0) return false;
        }
        return true;
    }

    void play() {
        cout << "\n🎴 === ГРА 'СІМ' РОЗПОЧАЛАСЯ! === 🎴\n\n";

        dealCards();

        while (true) {
            if (!makeMove()) {
                break;
            }

            if (allPlayersPassed()) {
                cout << "\nВсі гравці пропустили ход. Гра закінчена!\n";

                // Знаходимо гравця з найменшою кількістю карт
                int minCards = players[0]->getCardCount();
                int winner = 0;

                for (size_t i = 1; i < players.size(); i++) {
                    if (players[i]->getCardCount() < minCards) {
                        minCards = players[i]->getCardCount();
                        winner = i;
                    }
                }

                cout << "\n🏆 Переможець: " << players[winner]->getName()
                     << " (залишилось " << minCards << " карт) 🏆\n";
                break;
            }
        }

        cout << "\nДякуємо за гру!\n";
    }
};

int main() {
    // Встановлюємо локаль для підтримки українських символів
    setlocale(LC_ALL, "");

    cout << "╔════════════════════════════════════════╗\n";
    cout << "║       🎴 ГРА 'СІМ' 🎴                 ║\n";
    cout << "║                                        ║\n";
    cout << "║  Розробник: Сергій Щербаков           ║\n";
    cout << "║  Email: sergiyscherbakov@ukr.net      ║\n";
    cout << "║  Telegram: @s_help_2010               ║\n";
    cout << "╚════════════════════════════════════════╝\n\n";

    cout << "Оберіть режим гри:\n";
    cout << "1. Гра проти комп'ютера\n";
    cout << "2. Гра проти іншого гравця\n";
    cout << "Ваш вибір: ";

    int mode;
    cin >> mode;

    SevenGame game;

    if (mode == 1) {
        cout << "\nВведіть ваше ім'я: ";
        string playerName;
        cin >> playerName;

        game.addPlayer(new HumanPlayer(playerName));
        game.addPlayer(new ComputerPlayer("Комп'ютер"));
    } else if (mode == 2) {
        cout << "\nВведіть ім'я першого гравця: ";
        string player1Name;
        cin >> player1Name;

        cout << "Введіть ім'я другого гравця: ";
        string player2Name;
        cin >> player2Name;

        game.addPlayer(new HumanPlayer(player1Name));
        game.addPlayer(new HumanPlayer(player2Name));
    } else {
        cout << "Неправильний вибір!\n";
        return 1;
    }

    game.play();

    return 0;
}

File: python/seven_game.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Гра "Сім" - карткова гра для 2-4 гравців

Розробник: Сергій Щербаков
Email: sergiyscherbakov@ukr.net
Telegram: @s_help_2010
"""

import random
from typing import List, Dict, Tuple, Optional
from enum import Enum


class Suit(Enum):
    """Масті карт"""
    HEARTS = 0
    DIAMONDS = 1
    CLUBS = 2
    SPADES = 3


SUIT_SYMBOLS = {
    Suit.HEARTS: "♥",
    Suit.DIAMONDS: "♦",
    Suit.CLUBS: "♣",
    Suit.SPADES: "♠"
}

SUIT_NAMES = {
    Suit.HEARTS: "Черви",
    Suit.DIAMONDS: "Буби",
    Suit.CLUBS: "Хрести",
    Suit.SPADES: "Піки"
}


class Card:
    """Клас для представлення карти"""

    def __init__(self, rank: int, suit: Suit):
        self.rank = rank
        self.suit = suit

    def __str__(self) -> str:
        """Текстове представлення карти"""
        rank_str = {
            11: "J",
            12: "Q",
            13: "K",
            14: "A"
        }.get(self.rank, str(self.rank))

        return f"{rank_str}{SUIT_SYMBOLS[self.suit]}"

    def __repr__(self) -> str:
        return self.__str__()

    def __lt__(self, other):
        """Порівняння для сортування"""
        if self.suit.value != other.suit.value:
            return self.suit.value < other.suit.value
        return self.rank < other.rank


class Deck:
    """Клас колоди карт"""

    def __init__(self):
        """Створюємо колоду з 36 карт (від 6 до туза)"""
        self.cards = []
        for suit in Suit:
            for rank in range(6, 15):  # 6-10, J(11), Q(12), K(13), A(14)
                self.cards.append(Card(rank, suit))

    def shuffle(self):
        """Перемішуємо колоду"""
        random.shuffle(self.cards)

    def deal_cards(self, count: int) -> List[Card]:
        """Роздаємо карти"""
        dealt = []
        for _ in range(min(count, len(self.cards))):
            dealt.append(self.cards.pop())
        return dealt

    def is_empty(self) -> bool:
        """Перевірка чи колода порожня"""
        return len(self.cards) == 0


class Player:
    """Базовий клас гравця"""

    def __init__(self, name: str):
        self.name = name
        self.hand: List[Card] = []

    def add_cards(self, cards: List[Card]):
        """Додаємо карти до руки"""
        self.hand.extend(cards)
        self.sort_hand()

    def sort_hand(self):
        """Сортуємо карти в руці"""
        self.hand.sort()

    def show_hand(self):
        """Показуємо карти гравця"""
        print(f"\n{self.name} має {len(self.hand)} карт(и):")
        for i, card in enumerate(self.hand, 1):
            print(f"{i}. {card}  ", end="")
            if i % 9 == 0:
                print()
        print()

    def has_cards(self) -> bool:
        """Перевірка чи є карти у гравця"""
        return len(self.hand) > 0

    def get_card_count(self) -> int:
        """Кількість карт у гравця"""
        return len(self.hand)

    def get_valid_moves(self, table: Dict[Suit, Tuple[int, int]]) -> List[int]:
        """Отримуємо список можливих ходів"""
        valid = []

        for i, card in enumerate(self.hand):
            # Перевірка чи можна зіграти сімку
            if card.rank == 7:
                if card.suit not in table:
                    valid.append(i)
                    continue

            # Перевірка чи можна зіграти іншу карту
            if card.suit in table:
                min_rank, max_rank = table[card.suit]

                if card.rank == min_rank - 1 or card.rank == max_rank + 1:
                    valid.append(i)

        return valid

    def select_card(self, table: Dict[Suit, Tuple[int, int]], can_play_seven: bool) -> int:
        """Вибір карти для ходу (має бути перевизначений у підкласах)"""
        raise NotImplementedError

    def play_card(self, index: int) -> Optional[Card]:
        """Грає карту з руки"""
        if 0 <= index < len(self.hand):
            return self.hand.pop(index)
        return None


class HumanPlayer(Player):
    """Людський гравець"""

    def select_card(self, table: Dict[Suit, Tuple[int, int]], can_play_seven: bool) -> int:
        """Вибір карти людиною"""
        self.show_hand()

        valid_moves = self.get_valid_moves(table)

        if not valid_moves:
            print("У вас немає можливих ходів. Пропускаєте хід.")
            input("Натисніть Enter для продовження...")
            return -1

        print(f"Можливі ходи: {[i + 1 for i in valid_moves]}")

        while True:
            try:
                choice = input("Виберіть карту (номер) або 0 для пропуску ходу: ")
                choice = int(choice)

                if choice == 0:
                    return -1

                choice -= 1  # Перетворюємо в індекс масиву

                if choice in valid_moves:
                    return choice

                print("Неправильний вибір! Спробуйте ще раз.")
            except (ValueError, KeyboardInterrupt):
                print("Неправильний ввід! Введіть число.")


class ComputerPlayer(Player):
    """Комп'ютерний гравець"""

    def select_card(self, table: Dict[Suit, Tuple[int, int]], can_play_seven: bool) -> int:
        """Вибір карти комп'ютером"""
        valid_moves = self.get_valid_moves(table)

        if not valid_moves:
            print(f"{self.name} пропускає хід.")
            return -1

        # Проста стратегія: вибираємо випадковий хід
        selected_idx = random.choice(valid_moves)
        print(f"{self.name} грає карту: {self.hand[selected_idx]}")

        return selected_idx


class SevenGame:
    """Головний клас гри"""

    def __init__(self):
        self.players: List[Player] = []
        self.table: Dict[Suit, Tuple[int, int]] = {}
        self.current_player = 0
        self.consecutive_passes: List[int] = []

    def add_player(self, player: Player):
        """Додаємо гравця до гри"""
        self.players.append(player)
        self.consecutive_passes.append(0)

    def deal_cards(self):
        """Роздаємо карти"""
        deck = Deck()
        deck.shuffle()

        cards_per_player = 36 // len(self.players)

        for player in self.players:
            player.add_cards(deck.deal_cards(cards_per_player))

    def show_table(self):
        """Показуємо стан столу"""
        print("\n" + "=" * 50)
        print("СТІЛ".center(50))
        print("=" * 50)

        if not self.table:
            print("Стіл порожній. Грайте сімку для початку!".center(50))
        else:
            for suit in Suit:
                if suit in self.table:
                    min_rank, max_rank = self.table[suit]
                    cards = []
                    for rank in range(min_rank, max_rank + 1):
                        cards.append(str(Card(rank, suit)))

                    print(f"{SUIT_NAMES[suit]}: {' '.join(cards)}")

        print("=" * 50 + "\n")

    def make_move(self) -> bool:
        """Виконуємо хід поточного гравця"""
        player = self.players[self.current_player]

        print(f"\n>>> Хід гравця: {player.name} ({player.get_card_count()} карт) <<<")

        self.show_table()

        can_play_seven = len(self.table) < 4

        selected_card = player.select_card(self.table, can_play_seven)

        if selected_card == -1:
            self.consecutive_passes[self.current_player] += 1
            self.current_player = (self.current_player + 1) % len(self.players)
            return True

        card = player.play_card(selected_card)

        if card:
            # Оновлюємо стіл
            if card.suit not in self.table:
                # Нова масть на столі (має бути сімка)
                self.table[card.suit] = (card.rank, card.rank)
            else:
                # Розширюємо діапазон для існуючої масті
                min_rank, max_rank = self.table[card.suit]
                new_min = min(min_rank, card.rank)
                new_max = max(max_rank, card.rank)
                self.table[card.suit] = (new_min, new_max)

            self.consecutive_passes[self.current_player] = 0

            print(f"{player.name} зіграв карту: {card}")

            # Перевірка на перемогу
            if not player.has_cards():
                print(f"\n🎉 {player.name} ВИГРАВ! 🎉")
                return False

        self.current_player = (self.current_player + 1) % len(self.players)
        return True

    def all_players_passed(self) -> bool:
        """Перевірка чи всі гравці пропустили хід"""
        return all(passes > 0 for passes in self.consecutive_passes)

    def play(self):
        """Головний ігровий цикл"""
        print("\n🎴 === ГРА 'СІМ' РОЗПОЧАЛАСЯ! === 🎴\n")

        self.deal_cards()

        while True:
            if not self.make_move():
                break

            if self.all_players_passed():
                print("\nВсі гравці пропустили хід. Гра закінчена!")

                # Знаходимо гравця з найменшою кількістю карт
                min_cards = min(p.get_card_count() for p in self.players)
                winners = [p for p in self.players if p.get_card_count() == min_cards]

                if len(winners) == 1:
                    print(f"\n🏆 Переможець: {winners[0].name} (залишилось {min_cards} карт) 🏆")
                else:
                    print(f"\n🏆 Нічия між: {', '.join(w.name for w in winners)} 🏆")
                break

        print("\nДякуємо за гру!")


def print_header():
    """Виводимо заголовок гри"""
    print("╔" + "═" * 48 + "╗")
    print("║" + "🎴 ГРА 'СІМ' 🎴".center(48) + "║")
    print("║" + " " * 48 + "║")
    print("║" + "Розробник: Сергій Щербаков".center(48) + "║")
    print("║" + "Email: sergiyscherbakov@ukr.net".center(48) + "║")
    print("║" + "Telegram: @s_help_2010".center(48) + "║")
    print("╚" + "═" * 48 + "╝\n")


def main():
    """Головна функція програми"""
    print_header()

    print("Оберіть режим гри:")
    print("1. Гра проти комп'ютера")
    print("2. Гра проти іншого гравця")

    try:
        mode = int(input("Ваш вибір: "))
    except ValueError:
        print("Неправильний вибір!")
        return

    game = SevenGame()

    if mode == 1:
        player_name = input("\nВведіть ваше ім'я: ")
        game.add_player(HumanPlayer(player_name))
        game.add_player(ComputerPlayer("Комп'ютер"))
    elif mode == 2:
        player1_name = input("\nВведіть ім'я першого гравця: ")
        player2_name = input("Введіть ім'я другого гравця: ")

        game.add_player(HumanPlayer(player1_name))
        game.add_player(HumanPlayer(player2_name))
    else:
        print("Неправильний вибір!")
        return

    game.play()


if __name__ == "__main__":
    main()
